#!/usr/bin/env python
# -*- coding: utf-8 -*-

__author__ = "Evangelos A. Dimopoulos"
__copyright__ = "Copyright 2021, University of Oxford"
__email__ = "antonisdim41@gmail.com"
__license__ = "MIT"

import os

from scripts.utilities import REF_GENOME_TABLE, OUT_GENOME_TABLE


include: "rules/adapterremoval.smk"
include: "rules/qc.smk"
include: "rules/clustering.smk"
include: "rules/assembly.smk"
include: "rules/mlst.smk"
include: "rules/annotation.smk"
include: "rules/pangenome.smk"
include: "rules/amr.smk"


if os.path.isfile(REF_GENOME_TABLE):

    include: "rules/bt2_aln.smk"
    include: "rules/htsbox.smk"
    include: "rules/seq_stats.smk"
    include: "rules/trees.smk"
    include: "rules/snp_calling.smk"
    include: "rules/serotyping.smk"


if os.path.isfile(OUT_GENOME_TABLE):

    include: "rules/outgroup.smk"
    include: "rules/tree_stats.smk"
    include: "rules/trait_comparisons.smk"
    include: "rules/muttui.smk"


rule all:
    input:
        # "qc/fastqc_summary.tsv",
        # "assemblies.done",
        # "assemblies/assemblies_stats_ecoli.tsv",
        # "panaroo_ecoli/pangenome_merged/pan_genome_reference.fa",
        # "panaroo_ecoli/pangenome_merged/core_gene_alignment.aln",
        # "panaroo_ecoli/pangenome_merged/snps_core_gene_alignment.aln",
        # "clustering_ecoli/summary_clusters_mlst_ecoli_plus_contx.tsv",
        # expand(
        #     "seqs_ecoli/ecoli_cluster_{num}_seq_summary.tsv",
        #     num=[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20,
        #          21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35],
        # ),
        # expand(
        #     "trees_ecoli/ecoli_cluster_{num}_iq.treefile",
        #     num=[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20,
        #          21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35],
        # ),
        # expand(
        #     "trees_stats_ecoli/ecoli_cluster_{num}_heritability.tsv",
        #     num=[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20,
        #          21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35],
        # ),
        # expand(
        #     "trees_stats_ecoli/ecoli_cluster_{num}_assoc_idx.tsv",
        #     num=[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20,
        #          21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35],
        # ),
        # "seqs_ecoli/ecoli_cluster_1000_seq_summary.tsv",
        # expand("msa_ecoli/ecoli_population_{num}_chr_aln_nrec_snps.fasta", num=[26, 17, 24]),
        # expand(
        #     "trees_ecoli/ecoli_population_{num}_iq.treefile",
        #     num=[26, 17, 24],
        # ),
        # expand(
        #      "trees_ecoli/ecoli_population_{num}_bactdating.tree",
        #      num=[26, 17, 24],
        # ),
        expand(
            "timed_trees_ecoli/ecoli_population_{num}/timetree.nexus",
            num=[26, 17, 24],
        ),
        expand(
            "trees_stats_ecoli/ecoli_population_{num}_ace_tr_summary.tsv",
            num=[26, 17, 24],
        ),
        # "aux_files/github_r.done",
        # "trees_stats_ecoli/ecoli_cluster_tree_stats.tsv",
        # "trees_stats_ecoli/ecoli_population_21_total_host_links.tsv",
        # expand(
        #     "muttui_ecoli/ecoli_cluster_{num}/all_included_mutations.csv",
        #     num=[21, 23, 16, 19, 14, 2, 8, 10, 18, 13, 28, 20, 27, 24, 7, 3, 11, 29],
        # ),
        # expand(
        #     "muttui_ecoli/ecoli_cluster_{num}_label/all_included_mutations.csv",
        #     num=[21, 23, 16, 19, 14, 2, 8, 10, 18, 13, 28, 20, 27, 24, 7, 3, 11, 29],
        # ),
        # expand(
        #     "ariba_card_ecoli/ecoli_cluster_{num}_report.csv",
        #     num=[21, 23, 16, 19, 14, 2, 8, 10, 18, 13, 28, 20, 27, 24, 7, 3, 11, 29],
        # ),
        # expand(
        #     "ariba_vfdb_core_ecoli/ecoli_cluster_{num}_report.csv",
        #     num=[21, 23, 16, 19, 14, 2, 8, 10, 18, 13, 28, 20, 27, 24, 7, 3, 11, 29],
        # ),
        # "abricate_card_ecoli/ecoli_cluster_1000_report.tsv",
        # "abricate_vfdb_ecoli/ecoli_cluster_1000_report.tsv",
        # "ariba_card_ecoli/ecoli_population_21_report.csv",
        # "serotyping_ecoli/ecoli_population_21_sero_report.tsv",
        # "gatk_ecoli/ecoli_cluster_1000_SNP_filter_table.tsv",
        # "gatk_ecoli/ecoli_cluster_1000_sfs.pdf",
        # "panaroo_campylobacter/pangenome_merged/snps_core_gene_alignment.aln",
        # "panaroo_cjejuni/pangenome_merged/snps_core_gene_alignment.aln",
        # "clustering_cjejuni/summary_clusters_mlst_cjejuni_plus_contx.tsv",
        # "trees_cjejuni/cjejuni_cluster_1000_iq.treefile",
        # "trees_stats_cjejuni/cjejuni_cluster_tree_stats.tsv",
        # "seqs_cjejuni/cjejuni_cluster_1000_seq_summary.tsv",
        # expand(
        #     "trees_stats_{pathogen}/{pathogen}_cluster_1000_total_host_links.tsv",
        #     pathogen=['cjejuni']
        # ),
        # "trees_stats_cjejuni/cjejuni_cluster_1000_rec_swfst.tsv",
        # "trees_stats_cjejuni/cjejuni_cluster_1000_rec_tajima.tsv",
        # "trees_stats_cjejuni/cjejuni_cluster_1000_heritability.tsv",
        # expand("trees_stats_cjejuni/cjejuni_cluster_1000_{rec}_fst.tsv", rec=['nrec', 'rec']),
        # expand("trees_stats_cjejuni/cjejuni_cluster_1000_{rec}_pairfst.tsv", rec=['nrec', 'rec']),
        # expand("trees_stats_cjejuni/cjejuni_cluster_1000_{rec}_dxy.tsv", rec=['nrec', 'rec']),
        expand(
            "trees_stats_cjejuni/cjejuni_cluster_1000_{set}_cohort_hist.pdf",
            set=[
                "sb27",
                "host",
                "humans",
                "americans",
                "americans-time",
                "america-all",
                "narms-2019",
                "sb27-cali-narms",
                "sb27-rest-narms",
            ],
        ),
        # expand(
        #     "abricate_{db}_cjejuni/cjejuni_cluster_1000_report.tsv",
        #     db=['card', 'vfdb', 'argannot', 'megares', 'ncbi', 'plasmidfinder', 'resfinder']
        # ),
